clients/drcachesim/tracer/kcore_copy.cpp

```diff
@@ -409,9 +438,47 @@ kcore_copy_t::read_kallsyms()
             kcore_code_segments_num_++;
             modules_ = kernel_module;
             kernel_module = nullptr;
+        } else if (has_module && strcmp(module, BPF_JIT_MODULE_NAME) == 0 &&
+                   is_function_symbol(type)) {
+            bpf_jit_symbols.insert(addr);
         }
     }
     ASSERT(kernel_module == nullptr, "failed to find kernel module");
+
+    if (!bpf_jit_symbols.empty()) {
+        /* We copy a page size worth of contents after each bpf-related function symbol
+         * in an effort to make sure that the complete function is copied. This is
+         * similar to perf adding page size to the highest kernel symbol in its own
+         * kcore copy logic.
+         */
+#define PAGE_SIZE 4096
+        proc_module_t *bpf_module = nullptr;
+        for (auto it = bpf_jit_symbols.begin(); it != bpf_jit_symbols.end();) {
+            uint64_t addr = *it;
+            if (bpf_module == nullptr) {
+                bpf_module = (proc_module_t *)dr_global_alloc(sizeof(proc_module_t));
+                bpf_module->start = addr;
+                bpf_module->end = addr + PAGE_SIZE;
+                ++it;
+                continue;
+            }
+            if (bpf_module->end < addr) {
+                /* Just extend the last module region. */
+                bpf_module->end = addr + PAGE_SIZE;
```

Comment:
See comment above about perf doing something similar.

https://github.com/torvalds/linux/blob/6d0dc8559c847e2dcd66c5dd93dbab3d3d887ff5/tools/perf/util/symbol-elf.c#L2478

https://github.com/torvalds/linux/blob/6d0dc8559c847e2dcd66c5dd93dbab3d3d887ff5/tools/perf/util/symbol-elf.c#L2343

Reply:
I really don't understand this code.  Maybe we should discuss offline.  You have a list of function symbols.  You are creating proc_module_t ranges covering those symbols.  If a symbol's address is higher than the one you just looked at, you expand the prior one's range to cover the new one: but that could be a huge distance covered; it could include other symbols in between (your list is not sorted right?), yet you'll make new ranges for those others later even though you included them here.

Reply:
> your list is not sorted right?

It's sorted. Note that I'm using an std::set.

Reply:
> if (bpf_module->end < addr) {

This shouldve been >=. My bad.

Reply:
Fixed the if condition. Remaining logic should WAI because the set is sorted.